File: api/schemas/export.py
"""The export request on the wire.

Two sources, and they travel differently on purpose.

A **table** travels whole, in the request body. It has to: the `/compare` board
is assembled in the browser from live bank endpoints and has no server-side
identity, so there is no id to send. Sending the table also settles the scope
question for free -- the browser puts either the filtered, sorted, visible rows
or the entire table into `table`, and the server never has to know which, or
re-derive a filter it cannot see.

A **report** travels as an id. The server already holds it, so uploading its
body would only create a way for the exported file to disagree with the stored
one.

Cells arrive pre-formatted (`display`) alongside their raw `value`. See the
module docstring of `api/export/document.py` for why the browser is the one that
formats: the rules live in `UI/src/lib/cell-display.ts` and a Python copy would
drift.

**Nothing here caps anything.** No row limit, no cell-length limit. A user who
exports a 4,000-row table gets 4,000 rows.
"""

import uuid
from typing import Annotated, Literal

from pydantic import BaseModel, Field, model_validator

#: The formats a table can become. A report is prose, so it gets a subset --
#: see `ExportRequest` below.
TableFormat = Literal["csv", "xlsx", "pdf", "docx"]
DOCUMENT_FORMATS = frozenset({"pdf", "docx"})


class ExportCellIn(BaseModel):
    """One cell: the datum, and the text the screen showed for it."""

    value: str | float | bool | None = Field(
        default=None,
        description="The raw datum. XLSX writes this, so the column stays arithmetic.",
    )
    display: str = Field(
        default="",
        description=(
            "What the table drew, already formatted by `cellDisplayText`. PDF and "
            "DOCX write this, so the page reads exactly like the screen."
        ),
    )
    href: str = Field(default="", description="Link cells, and the source column.")
    note: str = Field(default="", description="The cell's hover note, if it has one.")
    tone: str = Field(
        default="",
        description="A badge cell's state: neutral | ok | warn | bad.",
    )


class ExportRowIn(BaseModel):
    """One row. `cells` is parallel to `columns` -- one entry each, in order."""

    cells: list[ExportCellIn]
    cite_url: str = Field(
        default="",
        description=(
            "Where this row came from. Becomes the appended source column: a rate "
            "without its source is a number nobody can check."
        ),
    )
    cite_note: str = Field(default="", description="Why that source supports the row.")


class Decimals(BaseModel):
    """A `number` column's precision, as `ResolvedColumn.decimals` resolved it."""

    min: int = 0
    max: int = 0


class ExportColumnIn(BaseModel):
    key: str
    label: str = ""
    type: str = Field(
        default="text",
        description="The contract's column type. Unknown values are treated as text.",
    )
    align: Literal["left", "center", "right"] = "left"
    currency: str = Field(default="", description="ISO code, for `money` columns.")
    decimals: Decimals | None = Field(
        default=None,
        description=(
            "Decimal places for a `number` column. Sent because Excel needs a "
            "number format, and an FX rate rounded to whole lira erases the very "
            "comparison the column exists for."
        ),
    )


class ExportTableIn(BaseModel):
    """A table as the user is looking at it."""

    title: str = ""
    subtitle: str = ""
    note: str = ""
    columns: list[ExportColumnIn]
    rows: list[ExportRowIn]

    @model_validator(mode="after")
    def _rows_match_columns(self) -> "ExportTableIn":
        width = len(self.columns)
        for index, row in enumerate(self.rows):
            if len(row.cells) != width:
                raise ValueError(
                    f"row {index} carries {len(row.cells)} cells for {width} columns"
                )
        return self


class TableSource(BaseModel):
    kind: Literal["table"] = "table"
    table: ExportTableIn


class ReportSource(BaseModel):
    kind: Literal["report"] = "report"
    report_id: uuid.UUID


Source = Annotated[TableSource | ReportSource, Field(discriminator="kind")]


class ExportRequest(BaseModel):
    """What to export, and as what."""

    format: TableFormat
    source: Source

    @model_validator(mode="after")
    def _format_suits_source(self) -> "ExportRequest":
        """A report is prose; a CSV of prose is a file nobody can open usefully.

        Refused here rather than in the router so the rule is visible in the
        OpenAPI schema and the frontend's generated types, and so the refusal is
        a 422 with a reason instead of a download that turns out to be useless.
        """
        if self.source.kind == "report" and self.format not in DOCUMENT_FORMATS:
            raise ValueError(
                "A report contains prose as well as tables, so it can only be "
                "exported as PDF or DOCX. Export a table from the report's own "
                "table controls to get CSV or XLSX."
            )
        return self

File: UI/src/lib/export.ts
/**
 * Turning what is on screen into the payload the export endpoint takes.
 *
 * The browser sends each cell **twice**: the raw datum and the text it drew.
 * That is not redundancy, it is the split that makes four formats possible from
 * one request. XLSX writes the datum, so a rate column can still be summed and
 * sorted in Excel; PDF and DOCX write the drawn text, so the page reads exactly
 * like the table it came from — `%2,89`, `₺1.234,56`, Turkish decimal comma and
 * all.
 *
 * Formatting happens here rather than in Python for the reason
 * `api/schemas/components.py` already states about prop schemas: the rules live
 * in TypeScript (`cell-display.ts`), and a second copy in another language would
 * drift the first time one side learned a new column type. The screen has
 * already answered "what does this cell say" in order to draw it; the export
 * asks for that answer rather than working it out again.
 *
 * **Nothing here caps anything.** Every row and every column the caller passes
 * goes into the payload whole.
 */

import type { components } from "@/types/api";

import { cellDisplayText, isBlankCell } from "./cell-display.ts";
import type { ResolvedColumn, Row } from "./contract";

type Schemas = components["schemas"];

export type ExportFormat = Schemas["ExportRequest"]["format"];
export type ExportTable = Schemas["ExportTableIn"];
export type ExportRequest = Schemas["ExportRequest"];

/** The order the dialog offers them in: cheapest and most portable first. */
export const TABLE_FORMATS: readonly ExportFormat[] = ["csv", "xlsx", "pdf", "docx"];

/**
 * What a report can become.
 *
 * A report is prose with tables in it, and a CSV of prose is a file nobody can
 * open usefully. The API refuses the other two with a reason (see
 * `api/schemas/export.py`); this list is why the dialog never offers them.
 */
export const REPORT_FORMATS: readonly ExportFormat[] = ["pdf", "docx"];

/**
 * One table, as the request body wants it.
 *
 * Give it the rows the user is *looking at*, or the whole table — that choice
 * belongs to the caller, and it is the only thing the scope toggle changes.
 * The server never learns which it got, because it never has to: there is no
 * filter state on that side to disagree with.
 */
export function tablePayload({
  columns,
  rows,
  title,
  subtitle = "",
  note = "",
  locale,
  bankLabels,
}: {
  columns: ResolvedColumn[];
  rows: Row[];
  title: string;
  subtitle?: string;
  note?: string;
  locale: "tr" | "en";
  bankLabels?: Record<string, string>;
}): ExportTable {
  return {
    title,
    subtitle,
    note,
    columns: columns.map((column) => ({
      key: column.key,
      label: column.label,
      type: column.type,
      align: column.align,
      currency: column.currency,
      decimals: column.decimals ?? null,
    })),
    rows: rows.map((row) => ({
      cite_url: row.cite_url ?? "",
      cite_note: row.cite_note ?? "",
      cells: columns.map((column) => {
        const value = row.cells[column.key];
        return {
          // `undefined` is not a JSON value; a cell the producer omitted has to
          // travel as an explicit null or the column shifts under it.
          value: isBlankCell(value) ? null : (value ?? null),
          display: isBlankCell(value)
            ? ""
            : cellDisplayText(value, column, locale, bankLabels),
          // A `link` cell's href is the value itself. `cellDisplayText` returns
          // the URL for those, but the two are separate fields on the wire
          // because the source column shows one and links to the other.
          href: column.type === "link" && typeof value === "string" ? value : "",
          note: row.cell_notes?.[column.key] ?? "",
          tone: row.cell_tones?.[column.key] ?? "",
        };
      }),
    })),
  };
}

/**
 * Hand a downloaded blob to the browser, under the name the server chose.
 *
 * The anchor has to be in the document before it is clicked — a detached one is
 * a no-op in Firefox — and the object URL has to be revoked afterwards or the
 * blob is pinned in memory for the life of the tab. A user exporting a dozen
 * comparison tables in a sitting is the case that turns into a leak.
 */
export function saveBlob(blob: Blob, filename: string): void {
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  anchor.rel = "noopener";
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}

/** `filename*=UTF-8''…`, falling back to `filename="…"`, falling back to a guess. */
export function filenameFrom(disposition: string | null, fallback: string): string {
  if (!disposition) return fallback;

  // RFC 5987 first: it is the one that carries Turkish letters, and every
  // current browser prefers it. `decodeURIComponent` can throw on malformed
  // input, and a bad header must not cost the user their download.
  const encoded = /filename\*=UTF-8''([^;]+)/i.exec(disposition);
  if (encoded) {
    try {
      return decodeURIComponent(encoded[1].trim());
    } catch {
      /* fall through to the plain parameter */
    }
  }

  const plain = /filename="([^"]+)"/i.exec(disposition);
  return plain ? plain[1] : fallback;
}
